File: tools/ml/installPythonModule.js
import { execSync } from 'child_process';

/**
 * Tool: Install Python Module
 * Installs a Python package using pip
 */
export default {
  name: 'installPythonModule',
  description: 'Install a Python package using pip',
  parameters: { moduleName: 'string' },
  
  async execute(params) {
    const { moduleName } = params;
    
    // Validate module name to prevent shell injection
    const moduleNameRegex = /^[a-zA-Z0-9._-]+$/;
    if (!moduleNameRegex.test(moduleName)) {
      return {
        success: false,
        message: `Invalid module name '${moduleName}'. Only alphanumeric characters, hyphens, underscores, and dots are allowed.`,
        error: 'Invalid module name'
      };
    }
    
    try {
      const output = execSync(`pip3 install ${moduleName}`, { 
        encoding: 'utf-8',
        timeout: 300000 // 5 minutes timeout
      });
      
      return {
        success: true,
        message: `Module '${moduleName}' installed successfully`,
        output: output
      };
    } catch (error) {
      return {
        success: false,
        message: `Failed to install module '${moduleName}': ${error.message}`,
        error: error.message
      };
    }
  }
};

File: obscurex/tools/ml/install_python_module.py
"""
Tool: Install Python Module
Installs a Python package using pip
"""

import subprocess
from typing import Dict, Any


async def execute(params: Dict[str, Any], context: Dict[str, Any] = None) -> Dict[str, Any]:
    """Install a Python package using pip."""
    module_name = params.get('moduleName')
    
    try:
        output = subprocess.run(
            ['pip3', 'install', module_name],
            capture_output=True,
            text=True,
            timeout=300,  # 5 minutes
            check=True
        )
        
        return {
            'success': True,
            'message': f"Successfully installed '{module_name}'",
            'output': output.stdout
        }
    except subprocess.TimeoutExpired:
        raise Exception(f"Installation of '{module_name}' timed out")
    except subprocess.CalledProcessError as e:
        raise Exception(f"Failed to install '{module_name}': {e.stderr}")
    except Exception as e:
        raise Exception(f"Failed to install '{module_name}': {str(e)}")


# Tool definition
tool = {
    'name': 'installPythonModule',
    'description': 'Install a Python package using pip',
    'parameters': {'moduleName': 'string'},
    'execute': execute
}
